pkg/odo/cli/component/create.go

```diff
@@ -263,8 +282,50 @@ func (co *CreateOptions) setResourceLimits() error {
 	return nil
 }
 
+// Validate if devfile component already exists
+func validateDevfileComponentCreateRequest() error {
+	if util.CheckPathExists(DevfilePath) {
+		return errors.New("Devfile.yaml already exists")
+	}
+
+	if util.CheckPathExists(EnvFilePath) {
+		return errors.New("env.yaml already exists")
+	}
+
+	return nil
+}
+
 // Complete completes create args
 func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	if experimental.IsExperimentalModeEnabled() {
+		co.devfileMetadata.componentType = args[0]
```

Comment:
This line is before the section that checks the number of args passed in to odo:

https://github.com/openshift/odo/blob/4852786c14ff643c490dbdd1ea31d6b699d1590f/pkg/odo/cli/component/create.go#L268-L270

If you run just `odo create`  without any args you'll get a panic from Go:
```
johns-mbp-3:odo johncollier$ ./odo create
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/openshift/odo/pkg/odo/cli/component.(*CreateOptions).Complete(0xc0001a23c0, 0xc00073c240, 0x6, 0xc000a02fc0, 0x3e2f790, 0x0, 0x0, 0x2, 0x279f0c7)
	/Users/johncollier/go/src/github.com/openshift/odo/pkg/odo/cli/component/create.go:301 +0x1b5b
```

We should:
- Only reference the args after we've validated the number of args passed in
- Support interactive mode if possible, as odo1 also supported it

Reply:
Also, how do we handle component name if user does not pass it in? I can do `odo create openLiberty` and no component name is set... we should fix that as well.

We should also check for the size of the component name, since it is going to create kube resources.

You can have a look at function https://github.com/openshift/odo/blob/master/pkg/odo/cli/component/devfile.go#L42 for trimming component name

I think this PR should also update the same line - https://github.com/openshift/odo/blob/master/pkg/odo/cli/component/devfile.go#L42, here we should read it from env.yaml, and you can test odo create & odo push with the same PR :D

Reply:
Good catch, I would go for the first option. I discussed the interactive mode work with Elson before, given finally we need to integrate odo2 with Codewind, CLI interactive mode doesn't have that much value for us and I believe our IDE can handle all interactive stuff properly.

Reply:
@maysunfaisal I remember last time Rajiv said `odo push` can pick up the component type as default component name if no component name is set, but to make it clear, I can make the env.yaml has the default component name if user doesn't specify.

If we already had the util function for sanitizing, I can directly apply it to `odo create` instead of `odo push` so that we don't need double sanitization.

That's all good suggestions!

Reply:
I think we should
- create component name in `odo create`
- if not provided during `odo create`, create one maybe like `devfile.go` `os.Getwd()` and store in `env.yaml`
- make sure component length is good, because we will create deployments, volume, url with component (check `devfile.go`)
- `odo push` should do `GetName()` from `envinf.go`. If not present push should error out. 

So `odo push` does not do component name creation, it just reads it from `env.yaml` and `odo create` does everything for the component name. I think current odo non-experimental is also like that? `odo create` creates component name in `config.yaml`

Reply:
@maysunfaisal Yes, that's what I am saying above, which `env.yaml` should have the component name field (either from user via `odo create` or the default component name we define) and we can reuse the sanitizing util function from `odo push` to sanitize (eg. string length, special character, etc). If we handle the component name on `odo create`, definitely that part is not necessary on `odo push`.

But for the default component name I would suggest we follow the same design as current odo which uses component type as default component name instead of using project directory path which the current `odo push` uses by `os.Getwd()` function. Reason is component type is less complicated than project directory path so that we can't reduce the burden of sanitizing.

The current odo handles all component generation configurations (config.yaml) on `odo create`, and `odo push` implements those configurations. We do the same for devfile support, only difference is generation configurations are devfile.yaml + env.yaml.

Reply:
@johnmcollier But in the meantime, I can take a look the interactive mode.